### ISSUE TITLE
Unable to use dynamic build property

### ISSUE THREAD

**Issue by wvanderdeijl** [pulumi/pulumi-docker#620]
### What happened?

When the `build` property to the `Image` resource is dynamic a `pulumi preview` fails since all command line arguments appear to be missing. This means our custom `foo.Dockerfile` is not being used and docker tries to use the default `Dockerfile` resulting in the following error:

```
Previewing update (skunkteam/dev)

View in Browser (Ctrl+O): https://app.pulumi.com/skunkteam/docker/dev/previews/053b8fb3-47c5-4c1c-956d-070192fa0953

     Type                        Name        Plan       Info
 +   pulumi:pulumi:Stack         docker-dev  create     
 +   ├─ random:index:RandomUuid  guid        create     
     └─ docker:index:Image       image                  1 error


Diagnostics:
  docker:index:Image (image):
    error: could not open dockerfile at relative path Dockerfile: stat Dockerfile: no such file or directory
```

### Expected Behavior

`pulumi preview` should not fail.

### Steps to reproduce

Full example at https://github.com/wvanderdeijl/docker-pulumi-issue

```typescript
import * as docker from "@pulumi/docker";
import { interpolate, output } from "@pulumi/pulumi";
import * as random from '@pulumi/random';

const apiKey = new random.RandomUuid('guid', {}).result

new docker.Image(
    'image',
    {
        imageName: 'eu.gcr.io/my-project/foo:v1.0.0',
        build: {
            dockerfile: interpolate`./foo.Dockerfile`,
            platform: 'linux/arm64',
            args: {
                KEY: apiKey,
            },
        },
        skipPush: true,
    },
);
```

### Output of `pulumi about`

CLI          
Version      3.65.1
Go Version   go1.20.3
Go Compiler  gc

Plugins
NAME    VERSION
docker  4.2.0
nodejs  unknown
random  4.13.0

Host     
OS       darwin
Version  13.3.1
Arch     arm64

This project is written in nodejs: executable='/Users/*****/.nvm/versions/node/v16.17.1/bin/node' version='v16.17.1'

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/*****
User           *****
Organizations  *****, *****

Dependencies:
NAME            VERSION
@pulumi/docker  4.2.0
@pulumi/pulumi  3.65.1
@pulumi/random  4.13.0
@types/node     16.18.25

Pulumi locates its logs in /var/folders/16/6r7yw0gd7593qv1z1zsp6xbh0000gn/T/ by default
warning: Failed to get information about the current stack: No current stack

### Additional context

When running a `pulumi preview` or `pulumi up` with `--logtostderr --logflow -v=9 2> out.txt` we can see in the logging that the `build` property is now a simple `output<string>` meaning the information for `dockerfile` (and `platform`) have been lost:
```
I0502 15:17:55.251320   89165 log.go:75] Unmarshaling property for RPC[ResourceMonitor.RegisterResource(docker:index/image:Image,image)]: build=output<string>{}
I0502 15:17:55.251336   89165 log.go:75] Unmarshaling property for RPC[ResourceMonitor.RegisterResource(docker:index/image:Image,image)]: imageName={eu.gcr.io/my-project/foo:v1.0.0}
I0502 15:17:55.251339   89165 log.go:75] Unmarshaling property for RPC[ResourceMonitor.RegisterResource(docker:index/image:Image,image)]: skipPush={true}
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by wvanderdeijl** [comment]
We ran into this issue when upgrading from `3.1.0` to 4`.2.0` so it seems that `3.1.0` did not suffer from this issue. I haven't tested versions in between those two.

**Comment by AaronFriel** [comment]
Hey @wvanderdeijl thanks for reporting this. I can suggest a workaround to try here:

```typescript
import * as docker from "@pulumi/docker";
import { interpolate, output, runtime } from "@pulumi/pulumi";
import * as random from '@pulumi/random';

const apiKey = new random.RandomUuid('guid', {}).result

if (!runtime.isDryRun()) {
    new docker.Image(
        'image',
        {
            imageName: 'eu.gcr.io/my-project/foo:v1.0.0',
            build: {
                dockerfile: interpolate`./foo.Dockerfile`,
                platform: 'linux/arm64',
                args: {
                    KEY: apiKey,
                },
            },
            skipPush: true,
        },
    );
}
```

This will skip evaluating the Docker.Image during previews, ensuring that the `apiKey` is known during the deployment ("up") phase. 

If this works for you, let us know - we may file a follow up issue to track improving preview behavior.

**Comment by wvanderdeijl** [comment]
That's a great tip. I'm now using that in a slightly different way. Still creating the `docker.Image` resource during preview but without the dynamic arguments. Otherwise the preview would show that the Image resource would be deleted. I've then added `{dependsOn:apiKey}` to explicitly define the dependency during preview.

**Comment by mikhailshilkov** [comment]
I closed #676 as a duplicate, but copying my notes here:

For engineers: The cause for it is that the `build` property comes as a single unknown

```
{"build":"04da6b54-80e4-46f7-96ec-b56ff0331ba9","imageName":"04da6b54-80e4-46f7-96ec-b56ff0331ba9","registry":{"password":"04da6b54-80e4-46f7-96ec-b56ff0331ba9","server":"04da6b54-80e4-46f7-96ec-b56ff0331ba9","username":"04da6b54-80e4-46f7-96ec-b56ff0331ba9"},"skipPush":false}
```

instead of a map with a few unknown properties (that would be similar to `registry above). In its turn, `build` is unknown because of the current implementation of codegen that sets the parent object as unknown if that object needs a default value (see [here](https://github.com/pulumi/pulumi-docker/blob/master/sdk/nodejs/image.ts#L177)).

We then discard unknowns [here](https://github.com/pulumi/pulumi-docker/blob/24161acd0433b061e3b5acfa128b9ddb599b965b/provider/provider.go#L131) and proceed with filling default build properties in [`marshalBuildAndApplyDefaults`](https://github.com/pulumi/pulumi-docker/blob/24161acd0433b061e3b5acfa128b9ddb599b965b/provider/provider.go#L139).
